Bottom: 6a0a4098a351b30b06ddfa0f29c7f02129cdcd09
Top:    f17cfac6aff2a45ea3bafc03fbc00980d439feb8
Author: Robert Collins <robertc@robertcollins.net>
Date:   2013-03-06 22:36:23 +1300

Refresh of extensible-run

---

diff --git a/testtools/run.py b/testtools/run.py
index 3e4765e..8dce5e4 100755
--- a/testtools/run.py
+++ b/testtools/run.py
@@ -13,6 +13,8 @@ import os
 import unittest
 import sys
 
+from extras import safe_hasattr
+
 from testtools import TextTestResult
 from testtools.compat import classtypes, istext, unicode_output_stream
 from testtools.testsuite import filter_by_ids, iterate_tests, sorted_tests
@@ -187,9 +189,9 @@ class TestProgram(object):
             self.runTests()
         else:
             runner = self._get_runner()
-            try:
+            if safe_hasattr(runner, 'list'):
                 runner.list(self.test)
-            except AttributeError:
+            else:
                 for test in iterate_tests(self.test):
                     stdout.write('%s\n' % test.id())
